M2/Macaulay2/packages/Permutations.m2

```diff
@@ -0,0 +1,383 @@
+    --     1. each cycle lists the largest element first and
+    --     2. the cycles are sorted by their first element in increasing order
+    sortedCycles := for cycle in cycles list toSequence rotate(maxPosition cycle, toList cycle);
+    sort sortedCycles
+)
+
+cycleType = method()
+cycleType Permutation := Sequence => w -> (
+    toSequence rsort for cycle in cycleDecomposition w list #cycle
+)
+
+------------------------------------
+-- Ascents, descents, runs, exceedances, and records
+-- NOTE: All return the 1-indexed results for consistency with the permutation notation
+------------------------------------
+ascents = method()
+ascents Permutation := List => w -> (
+    for i in 1 ..< #w list if w_i < w_(i+1) then i else continue
+)
+
+descents = method()
+descents Permutation := List => w -> (
+    for i in 1 ..< #w list if w_i > w_(i+1) then i else continue
+)
+
+ascendingRuns = method()
+ascendingRuns Permutation := List => w -> (
+    -- inspired from the SageMath implementation
+    -- https://github.com/sagemath/sage/blob/develop/src/sage/combinat/permutation.py
+    if #w == 1 then allRuns := {toSequence w}
+    else (
+        allRuns = {};
+        currentRun := {w#0};
+        for wi in w_(toList(1 ..< #w)) do (
+            if wi > last currentRun then currentRun = append(currentRun, wi)
+            else (
+                allRuns = append(allRuns, toSequence currentRun);
+                currentRun = {wi};
+            );
+        );
+        allRuns = append(allRuns, toSequence currentRun);
+    );
+    allRuns
+)
+
+descendingRuns = method()
+descendingRuns Permutation := List => w -> (
+    -- inspired from the SageMath implementation
+    -- https://github.com/sagemath/sage/blob/develop/src/sage/combinat/permutation.py
+    if #w == 1 then allRuns := {toSequence w}
+    else (
+        allRuns = {};
+        currentRun := {w#0};
+        for wi in w_(toList(1 ..< #w)) do (
+            if wi < last currentRun then currentRun = append(currentRun, wi)
+            else (
+                allRuns = append(allRuns, toSequence currentRun);
+                currentRun = {wi};
+            );
+        );
+        allRuns = append(allRuns, toSequence currentRun);
+    );
+    allRuns
+)
+
+exceedances = method(Options => {Weak => false})
+exceedances Permutation := List => opts -> w -> (
+    compare := if opts.Weak then ((i,j) -> i <= j) else ((i,j) -> i < j);
+    for i in 1 .. #w list if compare(i,w_i) then i else continue 
+)
+
+saliances = method()
+saliances Permutation := List => w -> (
+    to1Index positions(1 .. #w, i -> all(i+1 .. #w, j -> w_i > w_j))
+)
+
+records = method()
+records Permutation := List => w -> (
+    to1Index positions(1 .. #w, i -> all(1 ..< i, j -> w_j < w_i))
+)
+
+------------------------------------
+-- Pattern avoidance
+------------------------------------
+avoidsPattern = method(TypicalValue => Boolean)
+avoidsPattern (Permutation,List) := (w, pattern) -> (
```

Comment:
You should mention here that this code is borrowed from MatrixSchubert's `isPatternAvoiding` (and any others that may be in similar situation), and also get in touch with their authors to see how the two packages should interoperate (though that doesn't have to happen in this PR).

Reply:
Ah, I didn't think to mention it in the code itself because of my comment above.

> I should also mention that this package includes code for pattern avoidance and is taken directly from the `MatrixSchubert` package. The idea is that since pattern avoidance is inherent to permutations, it'll go into `Permutations`, and following the adoption of this package, the `MatrixSchubert` package might be overhauled to use the `Permutation` type from this package for consistency.

As one of the authors myself, I will be coordinating with Ayah to manage the overhaul for `MatrixSchubert`.

Reply:
Oh I didn't realize you are one of the authors. Never mind then haha.